File: server/router/rider.py
from fastapi import APIRouter 
from typing import List
from beanie import PydanticObjectId

from ..services.rider import RiderService
from ..models.rider import Rider
from ..dtos import RiderDTO

router = APIRouter(prefix="/rider", tags=["rider"])

@router.get("/{rider_id}", response_model=RiderDTO)
async def get_rider(rider_id: PydanticObjectId):
    return await RiderService.get_rider(rider_id)

@router.get("/", response_model=List[RiderDTO])
async def get_riders():
    return await RiderService.get_riders()

@router.post("/")
async def create_riders(riders: List[Rider]):
    return await RiderService.create_riders(riders)

@router.delete("/{rider_id}")
async def delete_rider(rider_id: PydanticObjectId):
    return await RiderService.delete_rider(rider_id) 

File: warehouse-optimization-client/src/store/api/rider.ts
import { riderAdapter } from "@/types/rider/adapter";
import { api } from "./base";
import { CreateRiderDTO, RiderDTO } from "@/types/rider/dto";
import { Rider } from "@/types/rider/type";
import _ from "lodash";	

const riderAPI = api.injectEndpoints({
	endpoints: (build) => ({
		getRiders: build.query<Rider[], void>({
			query: () => ({
				url: "rider/",
				method: "GET",
			}),
			providesTags: ["Rider"],
            transformResponse: (response: RiderDTO[]): Rider[] => {
                return _.map(response, riderAdapter);
            },
		}),

		getRider: build.query<Rider, string>({
			query: (riderId) => ({
				url: `rider/${riderId}`,
				method: "GET",
			}),
			providesTags: (result, error, id) => [{ type: "Rider", id }],
            transformResponse: (response: RiderDTO): Rider => {
                return riderAdapter(response);
            },
		}),

		createRiders: build.mutation<void, CreateRiderDTO[]>({
			query: (riders) => ({
				url: "rider/",
				method: "POST",
				body: riders,
			}),
			invalidatesTags: ["Rider"],
		}),

		deleteRider: build.mutation<void, string>({
			query: (riderId) => ({
				url: `rider/${riderId}`,
				method: "DELETE",
			}),
			invalidatesTags: (result, error, id) => [
				{ type: "Rider", id },
				"Rider",
				"DeliveryTasksBatch",
				"DeliveryTask",
			],
		}),
	}),
	overrideExisting: false,
});

export const {
	useGetRidersQuery,
	useGetRiderQuery,
	useCreateRidersMutation,
	useDeleteRiderMutation,
} = riderAPI;
